### PR TITLE
Diagnostic tool for Satellite 

### PR DESCRIPTION
**Pull request by aligeti**
https://storjlabs.atlassian.net/browse/V3-621

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -105,6 +105,7 @@ require (
 	github.com/go-ole/go-ole v1.2.1 // indirect
 	github.com/hanwen/go-fuse v0.0.0-20181011180456-b760b55765be
 	github.com/inconshreveable/mousetrap v1.0.0 // indirect
+	github.com/jedib0t/go-pretty v4.0.0+incompatible
```
Comment (egonelbre) [inline go.mod]: I think this isn't needed anymore.

File: go.sum [hunk 2]
```diff
@@ -153,6 +153,8 @@ github.com/inconshreveable/mousetrap v1.0.0 h1:Z8tu5sraLXCXIcARxBp/8cbvlwVa7Z1NH
 github.com/inconshreveable/mousetrap v1.0.0/go.mod h1:PxqpIevigyE2G7u3NXJIT2ANytuPF1OarO4DADm73n8=
 github.com/jbenet/go-base58 v0.0.0-20150317085156-6237cf65f3a6 h1:4zOlv2my+vf98jT1nQt4bT/yKWUImevYPJ2H344CloE=
 github.com/jbenet/go-base58 v0.0.0-20150317085156-6237cf65f3a6/go.mod h1:r/8JmuR0qjuCiEhAolkfvdZgmPiHTnJaG0UXCSeR1Zo=
+github.com/jedib0t/go-pretty v4.0.0+incompatible h1:3I/BQ59W64OTQUCSwW2f7lX5xqBWjkqpr6XQ80tymS4=
+github.com/jedib0t/go-pretty v4.0.0+incompatible/go.mod h1:XemHduiw8R651AF9Pt4FwCTKeG3oo7hrHJAoznj9nag=
```
Comment (egonelbre) [inline go.sum]: I think this isn't needed anymore.

File: pkg/bwagreement/server_test.go [hunk 3]
```diff
@@ -114,7 +118,7 @@ func NewTestServer(t *testing.T) *TestServer {
 }
 
 func newTestServerStruct(t *testing.T) *Server {
-	psqlInfo := getPSQLInfo()
+	psqlInfo := "postgres://postgres@localhost/pointerdb?sslmode=disable"
```
Comment (egonelbre) [inline pkg/bwagreement/server_test.go]: Please see how it's done in https://github.com/storj/storj/blob/master/storage/postgreskv/client_test.go#L27

File: pkg/bwagreement/server_test.go [hunk 4]
```diff
@@ -152,3 +156,22 @@ func (TS *TestServer) Stop() {
 	}
 	TS.grpcs.Stop()
 }
+
+// call this function to copy signature and data into postgres db
+func readSampleDataFromPsdb() (map[string][]*psdb.Agreement, error) {
```
Comment (egonelbre) [inline pkg/bwagreement/server_test.go]: Let's fill the database with actual data.

File: go.mod (hunk 1 above)
Comment (aligeti) [reply go.mod]: removed it

File: go.sum (hunk 2 above)
Comment (aligeti) [reply go.sum]: removed the references

File: pkg/bwagreement/server_test.go (hunk 3 above)
Comment (aligeti) [reply pkg/bwagreement/server_test.go]: I used visual code studio and you can run directly the test case with a mouse click and also implemented the support with a command line parameter as suggested.

File: pkg/bwagreement/server_test.go (hunk 4 above)
Comment (aligeti) [reply pkg/bwagreement/server_test.go]: Am bit not clear on the suggestion.

Below is the thought process of testing server.go's BandwidthAgreements()

Helper function readSampleDataFromPsdb() --> Implements reading from Piecestore's 'bandwidth_agreements' table and returns all the entries from that table as a slice. In TestBandwidthAgreements () test case I marshal the needed data(renter's signature and data) and send it through GRPC to satellite's bw agreement Postgres database ('pointerdb') and makes an entry into a table 'bwagreements'. This is test case tests server.go's BandwidthAgreements() which tests the creation of entry into Postgres table.

Comment (aligeti) [reply pkg/bwagreement/server_test.go]: yep I see your point...making mod

File: pkg/bwagreement/server_test.go [hunk 5]
```diff
@@ -113,9 +96,19 @@ func NewTestServer(t *testing.T) *TestServer {
...
+	testPostgres = flag.String("postgres-test-db", os.Getenv("STORJ_POSTGRESKV_TEST"), "PostgreSQL test database connection string")
+)
+
 func newTestServerStruct(t *testing.T) *Server {
-	psqlInfo := getPSQLInfo()
-	s, err := NewServer("postgres", psqlInfo, zap.NewNop())
+
+	if *testPostgres == "" {
+		//hard coded for local default, run locally --> psql -c 'create database pointerdb' -U postgres
+		*testPostgres = "postgres://postgres@localhost/pointerdb?sslmode=disable"
```
Comment (egonelbre) [inline pkg/bwagreement/server_test.go]: Do not hard code this, skip the test with an error. See the example how to do this. You may end up destroying someones local database this way.

File: pkg/bwagreement/server.go [hunk 6]
```diff
@@ -87,6 +100,7 @@ func (s *Server) BandwidthAgreements(stream pb.Bandwidth_BandwidthAgreementsServ
 			return nil
 		case agreement := <-ch:
 			go func() {
+				defer s.locked()()
```
Comment (egonelbre) [inline pkg/bwagreement/server.go]: Why use a goroutine when you are going to lock the whole server?

File: pkg/bwagreement/server_test.go [hunk 7]
```diff
@@ -1,15 +1,13 @@
-// Copyright (C) 2018 Storj Labs, Inc.
-// See LICENSE for copying information.
-
 package bwagreement
```
Comment (egonelbre) [inline pkg/bwagreement/server_test.go]: Copyright was removed

File: cmd/satellite/main.go [hunk 8]
```diff
@@ -127,6 +145,88 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
...
+		fmt.Println("reading bandwidth table error")
+		return err
+	}
+
+	const padding = 3
+	w := tabwriter.NewWriter(os.Stdout, 0, 0, padding, ' ', tabwriter.AlignRight|tabwriter.Debug)
+	fmt.Fprintln(w, "UplinkID\tTotal\t# Of Transactions\tPUT Action\tGET Action\t")
+
+	// Agreement is a struct that contains a bandwidth agreement and the associated signature
+	type UplinkAttributes struct {
```
Comment (egonelbre) [inline cmd/satellite/main.go]: `UplinkSummary`

File: cmd/satellite/main.go [hunk 9]
```diff
@@ -127,6 +145,88 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
...
+		TotalBytes        int64
+		PutActionCount    int64
+		GetActionCount    int64
+		TotalTransactions int64
+		// additional attributes add here ...
+	}
+
+	uplinkID := make(map[string]UplinkAttributes)
+	uplinkAtt := UplinkAttributes{}
+	var currUplinkID, lastUplinkID string
```
Comment (egonelbre) [inline cmd/satellite/main.go]: Same here as in the other issue. Don't use the stateful approach, lookup from the map for every entry.

File: cmd/satellite/main.go [hunk 10]
```diff
@@ -127,6 +145,88 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
...
+	// Agreement is a struct that contains a bandwidth agreement and the associated signature
+	type UplinkAttributes struct {
+		TotalBytes        int64
+		PutActionCount    int64
+		GetActionCount    int64
+		TotalTransactions int64
+		// additional attributes add here ...
+	}
+
+	uplinkID := make(map[string]UplinkAttributes)
```
Comment (egonelbre) [inline cmd/satellite/main.go]: `summaries`

File: cmd/satellite/main.go [hunk 11]
```diff
@@ -127,6 +145,88 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
...
+					} else {
+						satIDVal.GetActionCount++
+					}
+					satIDVal.TotalTransactions++
+					uplinkID[satIDKey] = satIDVal
+				}
+			}
+		}
+	}
+	fmt.Fprintln(w, currUplinkID, "\t", uplinkID[currUplinkID].TotalBytes, "\t", uplinkID[currUplinkID].TotalTransactions, "\t", uplinkID[currUplinkID].PutActionCount, "\t", uplinkID[currUplinkID].GetActionCount, "\t")
```
Comment (egonelbre) [inline cmd/satellite/main.go]: Why bother adding and getting them all when you only print one of them.

File: cmd/satellite/main.go [hunk 12]
```diff
@@ -127,6 +145,88 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
...
+		action := pbad.GetAction()
+
+		if strings.Compare(currUplinkID, lastUplinkID) != 0 {
+			// make an entry
+			uplinkAtt.TotalBytes = total
+			if action == pb.PayerBandwidthAllocation_PUT {
+				uplinkAtt.PutActionCount++
+			} else {
+				uplinkAtt.GetActionCount++
+				fmt.Println(uplinkAtt.GetActionCount)
```
Comment (egonelbre) [inline cmd/satellite/main.go]: Unnecessary print.

File: pkg/bwagreement/server_test.go (hunk 5 above)
Comment (aligeti) [reply pkg/bwagreement/server_test.go]: removed and made change as suggested

File: pkg/bwagreement/server.go (hunk 6 above)
Comment (aligeti) [reply pkg/bwagreement/server.go]: Done. removed the go fun()

File: pkg/bwagreement/server_test.go (hunk 7 above)
Comment (aligeti) [reply pkg/bwagreement/server_test.go]: Done

File: cmd/satellite/main.go (hunk 8 above)
Comment (aligeti) [reply cmd/satellite/main.go]: Made changes similar to the suggested in storagenode diag tool

File: cmd/satellite/main.go (hunk 9 above)
Comment (aligeti) [reply cmd/satellite/main.go]: Made changes similar to, as suggested in storagenode diag tool

File: cmd/satellite/main.go (hunk 10 above)
Comment (aligeti) [reply cmd/satellite/main.go]: Made changes similar to, as suggested in storagenode diag tool

File: cmd/satellite/main.go (hunk 12 above)
Comment (aligeti) [reply cmd/satellite/main.go]: Made changes similar to, as suggested in storagenode diag tool

File: go.sum (hunk 2 above)
Comment (egonelbre) [reply go.sum]: still here

File: go.mod (hunk 1 above)
Comment (egonelbre) [reply go.mod]: still here

File: cmd/satellite/main.go [hunk 13]
```diff
@@ -130,6 +148,82 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
...
+	dbpath := diagCfg.BasePath
+	s, err := bwagreement.NewServer("postgres", dbpath, zap.NewNop())
+	if err != nil {
+		fmt.Println("Storagenode database couldnt open:", dbpath)
+		return err
+	}
+	//get all bandwidth aggrements rows already ordered
+	baRows, err := s.GetBandwidthAllocations(context.Background())
+	if err != nil {
+		fmt.Println("satellite postgres 'bwagreements' table reading error:", dbpath)
```
Comment (egonelbre) [inline cmd/satellite/main.go]: `fmt.Printf("error reading satellite database %v: %v\n", dbpath, err)`

File: cmd/satellite/main.go [hunk 14]
```diff
@@ -130,6 +148,82 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
...
+		uplinkID := string(pbad.GetUplinkId())
+		summary, ok := summaries[uplinkID]
+		if !ok {
+			summaries[uplinkID] = &UplinkSummary{}
+			uplinkIDs = append(uplinkIDs, uplinkID)
+			summary = summaries[uplinkID]
+		}
+
+		// fill the summary info
+		summary.TotalBytes = summary.TotalBytes + rbad.GetTotal()
```
Comment (egonelbre) [inline cmd/satellite/main.go]: `summary.TotalBytes += rbad.GetTotal()`

File: pkg/bwagreement/server.go [hunk 15]
```diff
@@ -86,14 +100,20 @@ func (s *Server) BandwidthAgreements(stream pb.Bandwidth_BandwidthAgreementsServ
...
-				_, err = s.Create(ctx, agreement)
-				if err != nil {
-					s.logger.Error("DB entry creation Error", zap.Error(err))
-					errch <- err
-				}
-			}()
+			_, err = s.Create(ctx, agreement)
+			if err != nil {
+				s.logger.Error("DB entry creation Error", zap.Error(err))
+				errch <- err
```
Comment (egonelbre) [inline pkg/bwagreement/server.go]: you can directly return an error now.

File: go.mod (hunk 1 above)
Comment (aligeti) [reply go.mod]: Done

File: go.sum (hunk 2 above)
Comment (aligeti) [reply go.sum]: Done

File: cmd/satellite/main.go (hunk 14 above)
Comment (aligeti) [reply cmd/satellite/main.go]: Done

File: cmd/satellite/main.go (hunk 13 above)
Comment (aligeti) [reply cmd/satellite/main.go]: Done

File: pkg/bwagreement/server.go (hunk 15 above)
Comment (aligeti) [reply pkg/bwagreement/server.go]: Done

**Pull request merged**